### PR TITLE
Upgrade to ffmpeg 3.1

### PR DESCRIPTION
**Pull request by DavidBennettPIO**
I still need to test this but I think it's pretty well there.

Also I want to update the examples to the 3.1 API also.

Hopefully this will be ready to merge tomorrow, and the 3.2 upgrade next week.

### CODE REVIEW

**Comment by DavidBennettPIO** [comment]
I have updated the examples to use the new API as much as possible.

I was having some bugs with the internal ffmpeg changes to the AVCodecContext / AVCodecParameters changes in AVStream. If I used the new way it would give incorrect pts values in both decoding and encoding so I have kept this using AVStream.codec rather then allocating a new one.

Other then that everything seems to work fine... and seeing as the official examples seem to avoid this issue (transcode is the only example that reencodes more then one stream and uses the old API until 3.3) my guess it's user error on my part (documentation is near impossible to find for this case) or 3.1 had a bug rather then a issue with the linking.

Like last time, I'm going to create the 3.2 patch before merging as I might find some minor issues.

@ljubobratovicrelja Did you want to review the changes before I merge them?

Thanks,
David

**Comment by DavidBennettPIO** [comment]
@ljubobratovicrelja I also have the 3.2 upgrade ready to go, I could not see any errors with the 3.1 work so I believe it's good to merge.

Still want to work on the examples for 3.2 to see if I can get away from AVStream.codec but I'll probably work on that next week and then the 3.3 upgrade after that.

File: examples/encode_example.d [hunk 1]
```diff
@@ -68,26 +71,32 @@ void main(string[] args) {
...
         // only 1 audio and video stream
-        if( in_stream.codec.codec_type == AVMediaType.AVMEDIA_TYPE_VIDEO) {
+        if( in_stream.codecpar.codec_type == AVMediaType.AVMEDIA_TYPE_VIDEO) {
             input_video_stream = in_stream;
             input_video_stream_index = i;
-            input_video_codec_ctx = in_stream.codec;
-        } else if(in_stream.codec.codec_type == AVMediaType.AVMEDIA_TYPE_AUDIO){
+            input_video_codec_ctx = input_video_stream.codec; // TODO remove deprecation. currently needed for time_base in decoding internals
+            input_video_codec = avcodec_find_decoder(in_stream.codecpar.codec_id);
+//            input_video_codec_ctx = avcodec_alloc_context3(input_video_codec);
```
Comment (ljubobratovicrelja) [inline examples/encode_example.d]: Maybe remove commented code?

File: examples/encode_example.d [hunk 2]
```diff
@@ -44,8 +44,11 @@ void main(string[] args) {
     AVStream* input_video_stream;
     AVStream* input_audio_stream;
 
-    AVCodecContext* input_video_codec_ctx = null;
-    AVCodecContext* input_audio_codec_ctx = null;
+    AVCodec* input_video_codec;
+    AVCodec* input_audio_codec;
```
Comment (ljubobratovicrelja) [inline examples/encode_example.d:54]: Pointer default value is `null`, right? No explicit assignment needed?

File: examples/encode_example.d [hunk 3]
```diff
@@ -68,26 +71,32 @@ void main(string[] args) {
...
             input_video_stream = in_stream;
             input_video_stream_index = i;
-            input_video_codec_ctx = in_stream.codec;
-        } else if(in_stream.codec.codec_type == AVMediaType.AVMEDIA_TYPE_AUDIO){
+            input_video_codec_ctx = input_video_stream.codec; // TODO remove deprecation. currently needed for time_base in decoding internals
+            input_video_codec = avcodec_find_decoder(in_stream.codecpar.codec_id);
+//            input_video_codec_ctx = avcodec_alloc_context3(input_video_codec);
+//            if(avcodec_parameters_to_context(input_video_codec_ctx, in_stream.codecpar) < 0)
+//                throw new Exception("Could not set input context from input video codec parameters");
+        } else if(in_stream.codecpar.codec_type == AVMediaType.AVMEDIA_TYPE_AUDIO){
```
Comment (ljubobratovicrelja) [inline examples/encode_example.d]: codestyle nitpick - on some lines you have a space before open braces, on others you don't. Do you use some kind of formatter? If your willing, take care of that - these are examples after all, it'd be good to have them neat, IMHO. :)

File: examples/encode_example.d [hunk 4]
```diff
@@ -68,26 +71,32 @@ void main(string[] args) {
...
             input_audio_stream = in_stream;
             input_audio_stream_index = i;
-            input_audio_codec_ctx = in_stream.codec;
+            input_audio_codec_ctx = input_audio_stream.codec; // TODO remove deprecation. currently needed for time_base in decoding internals
+            input_audio_codec = avcodec_find_decoder(in_stream.codecpar.codec_id);
+//            input_audio_codec_ctx = avcodec_alloc_context3(input_audio_codec);
+//            if(avcodec_parameters_to_context(input_audio_codec_ctx, in_stream.codecpar) < 0)
+//                throw new Exception("Could not set input context from input audio codec parameters");
         } else continue;
```
Comment (ljubobratovicrelja) [inline examples/encode_example.d]: Maybe an unneeded empty line?

File: examples/encode_example.d [hunk 5]
```diff
@@ -177,32 +205,37 @@ void main(string[] args) {
...
     output_audio_codec_ctx.channel_layout   = av_get_default_channel_layout(input_audio_codec_ctx.channels);
     output_audio_codec_ctx.sample_rate      = input_audio_codec_ctx.sample_rate;
 //    output_audio_codec_ctx.sample_fmt       = input_audio_codec_ctx.sample_fmt;
     output_audio_codec_ctx.sample_fmt       = output_audio_codec.sample_fmts[0];
 
-//    output_audio_codec_ctx.channels         = 2;
-//    output_audio_codec_ctx.channel_layout   = av_get_default_channel_layout(2);
-//    output_audio_codec_ctx.sample_rate      = input_audio_codec_ctx.sample_rate;
-
-
```
Comment (ljubobratovicrelja) [inline examples/encode_example.d]: Why double newline? I'd say one is enough here.

**Review by ljubobratovicrelja — APPROVED** [review]
@complistic-gaff I've strolled through examples - only commented on some code style nits, but haven't checked the functionality. As I've said earlier, I'm trusting you on that one.

Anyways, stuff you're doing here is great so feel free to merge whenever you think its ready. I'm sorry for not going in depth with the review - changes in the library are not trivial, and unfortunately, I'm not familiar with ffmpeg 3.x. But as long as we're having different branches for each FFmpeg versions, we're safe for future fixes in each of them. So no change can be that bad, as I see it. :)

**Comment by DavidBennettPIO** [comment]
@ljubobratovicrelja I have updated the code style on the examples just recompiling my ffmpeg 3.1 now so I can test it. I'll let you know when thats done.

**Comment by DavidBennettPIO** [comment]
@ljubobratovicrelja looks like the examples work as expected, I'll look at merging this in a few hours.

**Pull request merged**